File: Sources/ChatView/LocalP2PTransport.swift
// Sources/ChatView/LocalP2PTransport.swift
//
// The `local-p2p` transport: a scripted, no-wire person-to-person / group backend that
// exercises the entire v2 vocabulary so the dual-alignment UI, the rows, paging, and the
// affordance gating all run end-to-end with no real transport. It is the demo content for
// the ChatPeople / ChatGroup example documents and the behavioral fixture source.
//
// Selected by `config.protocol == "local-p2p"`; `config.transport.scenario` picks "people"
// (a 1:1 chat, default) or "group" (four participants with member / call events). It
// advertises ALL capabilities, seeds a short recent conversation (timestamps, delivery
// statuses, a reaction, an edit, a tombstone, a file, a voice message), answers each user
// send with a scripted reply and a typing burst, honors reactions / edits / deletes /
// resends locally, and serves synthetic paged history on `.loadEarlier`.
//
// `@unchecked Sendable`: all mutable state is guarded by `lock`; the AsyncStream
// continuation is safe to yield from any context.

import Foundation
import CoreGraphics

final class LocalP2PTransport: ChatTransport, @unchecked Sendable {

    let events: AsyncStream<ChatEvent>
    private let continuation: AsyncStream<ChatEvent>.Continuation
    let capabilities = ChatTransportCapabilities(
        paging: true, typing: true, reactions: true, editing: true,
        deletion: true, replies: true, readReceipts: true, fileTransfer: true,
        messageIdentity: true, reportsConnectionState: true)

    private let scenario: String            // "people" | "group"
    private let stepDelay: UInt64           // ns between scripted beats (0 in tests)
    private let lock = NSLock()
    private var replyCounter = 0            // peer / generated / own server ids (peer-N, srv-N)
    private var tasks: [Task<Void, Never>] = []

    // Synthetic older history for paging: OLDEST first. `.loadEarlier` serves pages from the tail.
    private var history: [ChatItem] = []
    private var historyCursor = 0           // how many of `history` (from the end) have been served

    // The other party in the 1:1 chat / the primary correspondent in the group.
    private let peerID = "alex"
    private let peerName = "Alex"

    init(config: ChatTransportConfig) {
        self.scenario = config.string("scenario") ?? "people"
        let stepMs = config.int("stepMs") ?? 700
        self.stepDelay = UInt64(max(0, stepMs)) * 1_000_000
        var captured: AsyncStream<ChatEvent>.Continuation!
        self.events = AsyncStream(bufferingPolicy: .unbounded) { captured = $0 }
        self.continuation = captured
    }

    // MARK: - Lifecycle

    func start() async {
        continuation.yield(.sessionReady(sessionID: "local-p2p", configOptions: []))
        continuation.yield(.participantsChanged(roster()))
        // The link comes up: the composer starts connection-gated (the store defaults to .connecting)
        // and enables on this. The store seeds .connecting, so this is already a visible transition;
        // at stepMs 0 (tests) it lands deterministically.
        continuation.yield(.connectionStateChanged(.connected))
        buildHistory()
        for event in seedConversation() {
            continuation.yield(event)
        }
        // A little life after load: the peer types, then sends one more line. Only in the paced
        // (interactive demo) mode - at stepMs 0 (tests / fixtures) the seed is deterministic.
        guard stepDelay > 0 else {
            return
        }
        spawn { [weak self] in
            guard let self else { return }
            await self.pause(2)
            self.emitPeerTyping()
            await self.pause(1)
            let id = self.nextID(prefix: "peer")
            self.continuation.yield(.typingChanged(isTyping: false, senderID: self.peerID, senderName: self.peerName))
            self.continuation.yield(.messageReceived(self.peerMessage(id: id, text: "Anyway - talk later!",
                                                                      minutesAgo: 0)))
        }
    }

    func stop() async {
        lock.withLock { tasks }.forEach { $0.cancel() }
        continuation.finish()
    }

    // MARK: - Commands

    func send(_ command: ChatCommand) async {
        switch command {
        case .prompt(let text):
            handleUserSend(text: text, replyTo: nil, localID: nextID(prefix: "user"))   // legacy path; mint a handle
        case .sendMessage(let text, let replyTo, let localID):
            handleUserSend(text: text, replyTo: replyTo, localID: localID)

        case .toggleReaction(let itemID, let emoji, let add):
            // Reflect the toggle authoritatively (the store sent optimistic intent; we echo the result).
            let reactions = add ? [Reaction(emoji: emoji, count: 1, mine: true)] : []
            continuation.yield(.reactionsChanged(itemID: itemID, reactions: reactions))

        case .editMessage(let itemID, let newText):
            continuation.yield(.messageEdited(itemID: itemID, newText: newText, editedAt: Self.stamp(minutesAgo: 0)))

        case .deleteMessage(let itemID):
            continuation.yield(.messageDeleted(itemID: itemID))

        case .resendMessage(let itemID):
            spawn { [weak self] in
                guard let self else { return }
                self.continuation.yield(.messageStatusChanged(itemID: itemID, status: .sending))
                await self.pause(1)
                self.continuation.yield(.messageStatusWatermark(status: .delivered, upToItemID: itemID))
            }

        case .loadEarlier(_, let limit):
            servePage(limit: limit)

        case .cancelFileTransfer(let itemID):
            continuation.yield(.fileProgress(itemID: itemID, progress: nil, transferStatus: .cancelled))

        case .markRead, .setTyping, .cancel, .permissionResponse, .setConfigOption:
            break
        }
    }

    // MARK: - User send -> scripted reply

    private func handleUserSend(text: String, replyTo: String?, localID: String) {
        // The store already appended the optimistic local message keyed by `localID`. Confirm a
        // server id FIRST (the store re-keys the item), then drive the delivery ladder on the SERVER
        // id - exactly what a server-authoritative transport (SimpleX, Matrix) does: no reconstructed
        // "user-N" counting, own messages addressed only by the id the server assigned.
        let serverID = nextID(prefix: "srv")
        continuation.yield(.messageIDConfirmed(localID: localID, serverID: serverID))
        spawn { [weak self] in
            guard let self else { return }
            await self.pause(1)
            self.continuation.yield(.messageStatusChanged(itemID: serverID, status: .sent))
            await self.pause(1)
            self.continuation.yield(.messageStatusChanged(itemID: serverID, status: .delivered))
            self.emitPeerTyping()
            await self.pause(2)
            self.continuation.yield(.typingChanged(isTyping: false, senderID: self.peerID, senderName: self.peerName))
            let replyID = self.nextID(prefix: "peer")
            self.continuation.yield(.messageReceived(self.peerMessage(
                id: replyID, text: self.scriptedReply(to: text), minutesAgo: 0)))
            // The peer reads our message.
            await self.pause(1)
            self.continuation.yield(.messageStatusWatermark(status: .read, upToItemID: serverID))
        }
    }

    private func scriptedReply(to text: String) -> String {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.lowercased().contains("?") {
            return "Good question - let me check and get back to you."
        }
        return "Got it: \u{201C}\(trimmed.prefix(80))\u{201D}. Sounds good!"
    }

    // MARK: - Seeded conversation

    private func seedConversation() -> [ChatEvent] {
        var events: [ChatEvent] = []
        if scenario == "group" {
            events.append(.memberEvent(MemberEvent(id: "ev-join-1", timestamp: Self.stamp(daysAgo: 1, minute: 0),
                                                   kind: .joined, subjectName: "Sam")))
        }
        events.append(.messageReceived(peerMessage(id: "seed-1", text: "Hey! Are we still on for tomorrow?",
                                                   daysAgo: 1, minute: 1)))
        events.append(.messageReceived(peerMessage(id: "seed-2", text: "I found a great spot for lunch.",
                                                   daysAgo: 1, minute: 1,
                                                   reactions: [Reaction(emoji: "\u{1F44D}", count: 1, mine: true)])))
        events.append(.messageReceived(ownMessage(id: "seed-3", text: "Yes! Sounds perfect.",
                                                  daysAgo: 0, minute: 0, status: .read)))
        events.append(.messageReceived(ownMessage(id: "seed-4", text: "Let me double-check the time.",
                                                  daysAgo: 0, minute: 1, status: .delivered,
                                                  editedAt: Self.stamp(daysAgo: 0, minute: 2))))
        if scenario == "group" {
            events.append(.callEvent(CallEvent(id: "call-1", timestamp: Self.stamp(daysAgo: 0, minute: 3),
                                               kind: .missedIncoming, isVideo: true)))
            events.append(.memberEvent(MemberEvent(id: "ev-rename-1", timestamp: Self.stamp(daysAgo: 0, minute: 4),
                                                   kind: .renamed, actorName: "Alex", detail: "Weekend Trip")))
        }
        // A file and a voice message from the peer.
        events.append(.fileAdded(ChatFile(id: "file-1", role: .remote, senderID: peerID, senderName: peerName,
                                          timestamp: Self.stamp(daysAgo: 0, minute: 5), name: "itinerary.pdf",
                                          sizeBytes: 284_160, url: URL(string: "https://example.test/itinerary.pdf"),
                                          kind: .file, transferStatus: .completed)))
        events.append(.fileAdded(ChatFile(id: "voice-1", role: .remote, senderID: peerID, senderName: peerName,
                                          timestamp: Self.stamp(daysAgo: 0, minute: 6), name: "voice-message.m4a",
                                          sizeBytes: 51_200, url: URL(string: "https://example.test/voice.m4a"),
                                          kind: .voice, durationSeconds: 14, transferStatus: .completed)))
        return events
    }

    // MARK: - Synthetic paging

    private func buildHistory() {
        // 200 older messages, OLDEST first, alternating parties, a few days back.
        var items: [ChatItem] = []
        for index in 0..<200 {
            let mine = index % 2 == 1
            let id = "hist-\(index)"
            let day = 6 - (index / 40)   // spread across several older days for day separators
            let text = "Earlier message #\(index + 1)"
            if mine {
                items.append(.message(ownMessage(id: id, text: text, daysAgo: day, minute: index % 40, status: .read)))
            } else {
                items.append(.message(peerMessage(id: id, text: text, daysAgo: day, minute: index % 40)))
            }
        }
        lock.withLock {
            history = items
            historyCursor = 0
        }
    }

    private func servePage(limit: Int) {
        let (page, hasMore) = lock.withLock { () -> ([ChatItem], Bool) in
            let remaining = history.count - historyCursor
            guard remaining > 0 else {
                return ([], false)
            }
            let take = min(limit, remaining)
            // Serve the tail-most unseen slice, in chronological order.
            let upper = history.count - historyCursor
            let lower = upper - take
            let slice = Array(history[lower..<upper])
            historyCursor += take
            return (slice, historyCursor < history.count)
        }
        spawn { [weak self] in
            guard let self else { return }
            await self.pause(1)
            self.continuation.yield(.historyPage(items: page, hasMore: hasMore))
        }
    }

    // MARK: - Helpers

    private func roster() -> [Participant] {
        var people = [
            Participant(id: "me", name: "You", isSelf: true),
            Participant(id: peerID, name: peerName, avatarURL: nil),
        ]
        if scenario == "group" {
            people.append(Participant(id: "sam", name: "Sam"))
            people.append(Participant(id: "jordan", name: "Jordan"))
        }
        return people
    }

    private func emitPeerTyping() {
        continuation.yield(.typingChanged(isTyping: true, senderID: peerID, senderName: peerName))
    }

    private func peerMessage(id: String, text: String, daysAgo: Int = 0, minute: Int = 0, minutesAgo: Int? = nil,
                             reactions: [Reaction]? = nil) -> ChatMessage {
        ChatMessage(id: id, role: .remote, text: text, isStreaming: false, senderID: peerID, senderName: peerName,
                    timestamp: minutesAgo.map { Self.stamp(minutesAgo: $0) } ?? Self.stamp(daysAgo: daysAgo, minute: minute),
                    reactions: reactions)
    }

    private func ownMessage(id: String, text: String, daysAgo: Int = 0, minute: Int = 0, status: MessageStatus,
                            editedAt: String? = nil) -> ChatMessage {
        ChatMessage(id: id, role: .local, text: text, isStreaming: false, senderID: "me",
                    timestamp: Self.stamp(daysAgo: daysAgo, minute: minute), status: status, editedAt: editedAt)
    }

    private func nextID(prefix: String) -> String {
        lock.withLock {
            replyCounter += 1
            return "\(prefix)-\(replyCounter)"
        }
    }

    private func spawn(_ body: @escaping @Sendable () async -> Void) {
        let task = Task(operation: body)
        lock.withLock { tasks.append(task) }
    }

    private func pause(_ beats: Int) async {
        guard stepDelay > 0 else { return }
        try? await Task.sleep(nanoseconds: stepDelay * UInt64(beats))
    }

    // A stable clock for the scripted timestamps: a fixed "now" so the seeded conversation is
    // deterministic (the demo shows dated day separators; tests get reproducible times).
    nonisolated(unsafe) private static let base: Date = ISO8601DateFormatter().date(from: "2026-07-10T15:00:00Z")!

    private static func stamp(daysAgo: Int = 0, minute: Int = 0, minutesAgo: Int? = nil) -> String {
        let seconds: TimeInterval
        if let minutesAgo {
            seconds = -Double(minutesAgo) * 60
        } else {
            seconds = -Double(daysAgo) * 86_400 + Double(minute) * 60
        }
        let date = base.addingTimeInterval(seconds)
        return isoFormatter.string(from: date)
    }

    nonisolated(unsafe) private static let isoFormatter: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]
        return formatter
    }()
}

File: android/chatview/src/main/kotlin/com/abracode/chatview/LocalP2PTransport.kt
package com.abracode.chatview

// Port of Sources/ChatView/LocalP2PTransport.swift.
//
// The `local-p2p` transport: a scripted, no-wire person-to-person / group backend that exercises the entire v2
// vocabulary so the dual-alignment UI, the rows, paging, and the affordance gating all run end-to-end with no real
// transport. It is the demo content for the ChatPeople / ChatGroup examples and the behavioral fixture source.
//
// `config.transport.scenario` picks "people" (a 1:1 chat, default) or "group" (four participants with member / call
// events). It advertises ALL capabilities, seeds a short recent conversation, answers each user send with a scripted
// reply and a typing burst, honors reactions / edits / deletes / resends locally, and serves synthetic paged
// history on `.loadEarlier`. Swift's AsyncStream + Task map to a Channel + coroutines on an injectable-dispatcher
// scope; mutable state is guarded like Swift's NSLock.

import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.channels.ReceiveChannel
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import java.time.Instant
import java.time.format.DateTimeFormatter

class LocalP2PTransport(
    config: ChatTransportConfig,
    dispatcher: CoroutineDispatcher = Dispatchers.Default,
) : ChatTransport {

    private val channel = Channel<ChatEvent>(Channel.UNLIMITED)
    override val events: ReceiveChannel<ChatEvent> get() = channel

    override val capabilities = ChatTransportCapabilities(
        paging = true, typing = true, reactions = true, editing = true,
        deletion = true, replies = true, readReceipts = true, fileTransfer = true,
        messageIdentity = true, reportsConnectionState = true,
    )

    private val scenario: String = config.string("scenario") ?: "people"   // "people" | "group"
    private val stepMs: Int = (config.int("stepMs") ?: 700).coerceAtLeast(0)  // ms between scripted beats (0 in tests)

    private val scope = CoroutineScope(SupervisorJob() + dispatcher)
    private val lock = Any()
    private var replyCounter = 0                        // peer / generated / own server ids (peer-N, srv-N)
    private val jobs = mutableListOf<Job>()

    // Synthetic older history for paging: OLDEST first. `.loadEarlier` serves pages from the tail.
    private var history: List<ChatItem> = emptyList()
    private var historyCursor = 0                       // how many of `history` (from the end) have been served

    private val peerID = "alex"
    private val peerName = "Alex"

    // Built from code points so the source stays ASCII: U+1F44D thumbs-up, U+201C / U+201D curly double quotes.
    private val thumbsUp = String(Character.toChars(0x1F44D))
    private val leftQuote = String(Character.toChars(0x201C))
    private val rightQuote = String(Character.toChars(0x201D))

    // MARK: - Lifecycle

    override suspend fun start() {
        emit(ChatEvent.SessionReady(sessionID = "local-p2p", configOptions = emptyList()))
        emit(ChatEvent.ParticipantsChanged(roster()))
        // The link comes up: the composer starts connection-gated (the store defaults to .connecting) and enables on
        // this. At stepMs 0 (tests) it lands deterministically.
        emit(ChatEvent.ConnectionStateChanged(ChatConnectionState.CONNECTED))
        buildHistory()
        for (event in seedConversation()) {
            emit(event)
        }
        // A little life after load: the peer types, then sends one more line. Paced (interactive demo) mode only - at
        // stepMs 0 (tests / fixtures) the seed is deterministic.
        if (stepMs <= 0) {
            return
        }
        spawn {
            pause(2)
            emitPeerTyping()
            pause(1)
            val id = nextID("peer")
            emit(ChatEvent.TypingChanged(isTyping = false, senderID = peerID, senderName = peerName))
            emit(ChatEvent.MessageReceived(peerMessage(id = id, text = "Anyway - talk later!", minutesAgo = 0)))
        }
    }

    override suspend fun stop() {
        synchronized(lock) { jobs.toList() }.forEach { it.cancel() }
        channel.close()
        scope.cancel()
    }

    // MARK: - Commands

    override suspend fun send(command: ChatCommand) {
        when (command) {
            is ChatCommand.Prompt ->
                handleUserSend(text = command.text, replyTo = null, localID = nextID("user"))  // legacy path; mint a handle
            is ChatCommand.SendMessage ->
                handleUserSend(text = command.text, replyTo = command.replyTo, localID = command.localID)

            is ChatCommand.ToggleReaction -> {
                // Reflect the toggle authoritatively (the store sent optimistic intent; we echo the result).
                val reactions = if (command.add) listOf(Reaction(emoji = command.emoji, count = 1, mine = true)) else emptyList()
                emit(ChatEvent.ReactionsChanged(itemID = command.itemID, reactions = reactions))
            }

            is ChatCommand.EditMessage ->
                emit(ChatEvent.MessageEdited(itemID = command.itemID, newText = command.newText, editedAt = stamp(minutesAgo = 0)))

            is ChatCommand.DeleteMessage ->
                emit(ChatEvent.MessageDeleted(itemID = command.itemID))

            is ChatCommand.ResendMessage -> spawn {
                emit(ChatEvent.MessageStatusChanged(itemID = command.itemID, status = MessageStatus.SENDING))
                pause(1)
                emit(ChatEvent.MessageStatusWatermark(status = MessageStatus.DELIVERED, upToItemID = command.itemID))
            }

            is ChatCommand.LoadEarlier -> servePage(command.limit)

            is ChatCommand.CancelFileTransfer ->
                emit(ChatEvent.FileProgress(itemID = command.itemID, progress = null, transferStatus = FileTransferStatus.CANCELLED))

            is ChatCommand.MarkRead, is ChatCommand.SetTyping, is ChatCommand.Cancel,
            is ChatCommand.PermissionResponse, is ChatCommand.SetConfigOption -> Unit
        }
    }

    // MARK: - User send -> scripted reply

    private fun handleUserSend(text: String, replyTo: String?, localID: String) {
        // The store already appended the optimistic local message keyed by `localID`. Confirm a server id FIRST (the
        // store re-keys the item), then drive the delivery ladder on the SERVER id - exactly what a
        // server-authoritative transport does: own messages addressed only by the id the server assigned.
        val serverID = nextID("srv")
        emit(ChatEvent.MessageIDConfirmed(localID = localID, serverID = serverID))
        spawn {
            pause(1)
            emit(ChatEvent.MessageStatusChanged(itemID = serverID, status = MessageStatus.SENT))
            pause(1)
            emit(ChatEvent.MessageStatusChanged(itemID = serverID, status = MessageStatus.DELIVERED))
            emitPeerTyping()
            pause(2)
            emit(ChatEvent.TypingChanged(isTyping = false, senderID = peerID, senderName = peerName))
            val replyID = nextID("peer")
            emit(ChatEvent.MessageReceived(peerMessage(id = replyID, text = scriptedReply(text), minutesAgo = 0)))
            // The peer reads our message.
            pause(1)
            emit(ChatEvent.MessageStatusWatermark(status = MessageStatus.READ, upToItemID = serverID))
        }
    }

    private fun scriptedReply(text: String): String {
        val trimmed = text.trim()
        if (trimmed.lowercase().contains("?")) {
            return "Good question - let me check and get back to you."
        }
        return "Got it: $leftQuote${trimmed.take(80)}$rightQuote. Sounds good!"
    }

    // MARK: - Seeded conversation

    private fun seedConversation(): List<ChatEvent> {
        val events = mutableListOf<ChatEvent>()
        if (scenario == "group") {
            events.add(
                ChatEvent.MemberEventReceived(
                    MemberEvent(id = "ev-join-1", timestamp = stamp(daysAgo = 1, minute = 0), kind = MemberEvent.Kind.JOINED, subjectName = "Sam"),
                ),
            )
        }
        events.add(ChatEvent.MessageReceived(peerMessage(id = "seed-1", text = "Hey! Are we still on for tomorrow?", daysAgo = 1, minute = 1)))
        events.add(
            ChatEvent.MessageReceived(
                peerMessage(
                    id = "seed-2", text = "I found a great spot for lunch.", daysAgo = 1, minute = 1,
                    reactions = listOf(Reaction(emoji = thumbsUp, count = 1, mine = true)),
                ),
            ),
        )
        events.add(ChatEvent.MessageReceived(ownMessage(id = "seed-3", text = "Yes! Sounds perfect.", daysAgo = 0, minute = 0, status = MessageStatus.READ)))
        events.add(
            ChatEvent.MessageReceived(
                ownMessage(
                    id = "seed-4", text = "Let me double-check the time.", daysAgo = 0, minute = 1,
                    status = MessageStatus.DELIVERED, editedAt = stamp(daysAgo = 0, minute = 2),
                ),
            ),
        )
        if (scenario == "group") {
            events.add(
                ChatEvent.CallEventReceived(
                    CallEvent(id = "call-1", timestamp = stamp(daysAgo = 0, minute = 3), kind = CallEvent.Kind.MISSED_INCOMING, isVideo = true),
                ),
            )
            events.add(
                ChatEvent.MemberEventReceived(
                    MemberEvent(id = "ev-rename-1", timestamp = stamp(daysAgo = 0, minute = 4), kind = MemberEvent.Kind.RENAMED, actorName = "Alex", detail = "Weekend Trip"),
                ),
            )
        }
        // A file and a voice message from the peer.
        events.add(
            ChatEvent.FileAdded(
                ChatFile(
                    id = "file-1", role = ChatRole.REMOTE, senderID = peerID, senderName = peerName,
                    timestamp = stamp(daysAgo = 0, minute = 5), name = "itinerary.pdf",
                    sizeBytes = 284_160, url = "https://example.test/itinerary.pdf",
                    kind = ChatFile.Kind.FILE, transferStatus = FileTransferStatus.COMPLETED,
                ),
            ),
        )
        events.add(
            ChatEvent.FileAdded(
                ChatFile(
                    id = "voice-1", role = ChatRole.REMOTE, senderID = peerID, senderName = peerName,
                    timestamp = stamp(daysAgo = 0, minute = 6), name = "voice-message.m4a",
                    sizeBytes = 51_200, url = "https://example.test/voice.m4a",
                    kind = ChatFile.Kind.VOICE, durationSeconds = 14, transferStatus = FileTransferStatus.COMPLETED,
                ),
            ),
        )
        return events
    }

    // MARK: - Synthetic paging

    private fun buildHistory() {
        // 200 older messages, OLDEST first, alternating parties, a few days back.
        val items = mutableListOf<ChatItem>()
        for (index in 0 until 200) {
            val mine = index % 2 == 1
            val id = "hist-$index"
            val day = 6 - (index / 40)   // spread across several older days for day separators
            val text = "Earlier message #${index + 1}"
            if (mine) {
                items.add(ChatItem.Message(ownMessage(id = id, text = text, daysAgo = day, minute = index % 40, status = MessageStatus.READ)))
            } else {
                items.add(ChatItem.Message(peerMessage(id = id, text = text, daysAgo = day, minute = index % 40)))
            }
        }
        synchronized(lock) {
            history = items
            historyCursor = 0
        }
    }

    private fun servePage(limit: Int) {
        val page: List<ChatItem>
        val hasMore: Boolean
        synchronized(lock) {
            val remaining = history.size - historyCursor
            if (remaining <= 0) {
                page = emptyList()
                hasMore = false
            } else {
                val take = minOf(limit, remaining)
                // Serve the tail-most unseen slice, in chronological order.
                val upper = history.size - historyCursor
                val lower = upper - take
                page = history.subList(lower, upper).toList()
                historyCursor += take
                hasMore = historyCursor < history.size
            }
        }
        spawn {
            pause(1)
            emit(ChatEvent.HistoryPage(items = page, hasMore = hasMore))
        }
    }

    // MARK: - Helpers

    private fun roster(): List<Participant> {
        val people = mutableListOf(
            Participant(id = "me", name = "You", isSelf = true),
            Participant(id = peerID, name = peerName, avatarURL = null),
        )
        if (scenario == "group") {
            people.add(Participant(id = "sam", name = "Sam"))
            people.add(Participant(id = "jordan", name = "Jordan"))
        }
        return people
    }

    private fun emitPeerTyping() {
        emit(ChatEvent.TypingChanged(isTyping = true, senderID = peerID, senderName = peerName))
    }

    private fun peerMessage(
        id: String,
        text: String,
        daysAgo: Int = 0,
        minute: Int = 0,
        minutesAgo: Int? = null,
        reactions: List<Reaction>? = null,
    ): ChatMessage = ChatMessage(
        id = id, role = ChatRole.REMOTE, text = text, isStreaming = false, senderID = peerID, senderName = peerName,
        timestamp = if (minutesAgo != null) stamp(minutesAgo = minutesAgo) else stamp(daysAgo = daysAgo, minute = minute),
        reactions = reactions,
    )

    private fun ownMessage(
        id: String,
        text: String,
        daysAgo: Int = 0,
        minute: Int = 0,
        status: MessageStatus,
        editedAt: String? = null,
    ): ChatMessage = ChatMessage(
        id = id, role = ChatRole.LOCAL, text = text, isStreaming = false, senderID = "me",
        timestamp = stamp(daysAgo = daysAgo, minute = minute), status = status, editedAt = editedAt,
    )

    private fun nextID(prefix: String): String = synchronized(lock) { "$prefix-${++replyCounter}" }

    private fun spawn(body: suspend () -> Unit) {
        val job = scope.launch { body() }
        synchronized(lock) { jobs.add(job) }
    }

    private suspend fun pause(beats: Int) {
        if (stepMs > 0) {
            delay(stepMs.toLong() * beats)
        }
    }

    private fun emit(event: ChatEvent) {
        channel.trySend(event)
    }

    companion object {
        // A stable clock for the scripted timestamps: a fixed "now" so the seeded conversation is deterministic.
        private val base: Instant = Instant.parse("2026-07-10T15:00:00Z")

        private fun stamp(daysAgo: Int = 0, minute: Int = 0, minutesAgo: Int? = null): String {
            val seconds: Long = if (minutesAgo != null) {
                -minutesAgo.toLong() * 60
            } else {
                -daysAgo.toLong() * 86_400 + minute.toLong() * 60
            }
            return DateTimeFormatter.ISO_INSTANT.format(base.plusSeconds(seconds))
        }
    }
}
